torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,215 @@
+      it->second.insert(name);
+    } else {
+      std::set<std::string> names;
+      names.insert(name);
+      preservedAttrs_[&module] = names;
+    }
+  }
+
+  // We can use a more efficient algorithm to hash each constant GetAttr to its
+  // conrresponding Value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
+      Node* input,
+      std::string m_name,
+      script::Module& attr_module) {
+    std::stack<std::string> names;
+
+    while (!attr_module.hasattr(m_name)) {
+      if (input->kind() == prim::GetAttr) {
+        m_name = input->s(attr::name);
+        names.push(m_name);
+        input = input->inputs()[0]->node();
+      } else {
+        // TODO: handle TupleUnpack.
+        return false;
+      }
+    }
+
+    while (!names.empty()) {
+      auto m_name = names.top();
+      names.pop();
+      attr_module = attr_module.attr(m_name).toModule();
+      auto it = preservedAttrs_.find(&attr_module);
+      if (it != preservedAttrs_.end()) {
+        if (it->second.count(m_name)) {
+          // Bail because this attribute is mutable!
+          return false;
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(&attr_module);
+    return it == preservedAttrs_.end() || !it->second.count(m_name);
+  }
+
+  void recordMutableAttrs(
+      std::shared_ptr<Graph>& graph,
+      const script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr) {
+          auto name = n->s(attr::name);
+          auto attr_module = module_clone;
+          auto input = n->inputs()[0]->node();
+          if (!findConstantAttr(input, name, attr_module)) {
+            continue;
+          }
+          insertModifiedAttr(attr_module, name);
+        }
+      }
+    }
+  }
+
+  std::set<std::string> getReferencedAttrs(
+      std::shared_ptr<Graph>& graph,
+      script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          if (module_clone.hasattr(name)) {
+            insertModifiedAttr(module_clone, name);
+          }
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(&module_clone);
+    if (it != preservedAttrs_.end())
+      return it->second;
+    else
+      return std::set<std::string>();
+  }
+
+ public:
+  void propagateAttributes(
```

Comment:
Do you expect users of this class to ever invoke only `propagateAttributes` or `cleanupFrozenModule`? If they are only expected to be called together, it would make sense to make them private and expose only a wrapper (a `run` method) - the code that currently lives in `FreezeModule`.